prov/util/src/util_mem_hooks.c

```diff
@@ -125,246 +154,209 @@ static struct ofi_intercept intercepts[] = {
-		struct ofi_dl_intercept, dl_entry, entry) {
-
-		if (dl_entry->dl_func_addr != func_addr)
-			continue;
-
-		assert(*func_addr == intercept->our_func);
-		*func_addr = dl_entry->dl_func;
-		dlist_remove(&dl_entry->entry);
-		free(dl_entry);
-		FI_DBG(&core_prov, FI_LOG_MR,
-		       "dl symbol %s restored\n", intercept->symbol);
-		break;
-	}
+static void ofi_restore_intercepts(void)
+{
+	struct ofi_intercept *intercept;
 
-	return FI_SUCCESS;
+	dlist_foreach_container(&memhooks.intercept_list, struct ofi_intercept,
+		intercept, entry)
+		ofi_remove_patch(intercept);
 }
 
-static int ofi_restore_phdr_handler(struct dl_phdr_info *info,
-                                    size_t size, void *data)
+#if (defined(__x86_64___) || defined(__amd64__))
+static int ofi_patch_function(struct ofi_intercept *intercept)
+{
+	intercept->patch_data_size = 13;
+        *(unsigned short*)(intercept->patch_data + 0) = 0xbb49;
+        *(unsigned long* )(intercept->patch_data + 2) =
+		(unsigned long) intercept->our_func;
+        *(unsigned char*) (intercept->patch_data +10) = 0x41;
+        *(unsigned char*) (intercept->patch_data +11) = 0xff;
+        *(unsigned char*) (intercept->patch_data +12) = 0xe3;
+
+	return ofi_apply_patch(intercept);
+}
+#elif defined(__aarch64__)
+/**
+ * @brief Generate a mov immediate instruction
+ *
+ * @param[in] reg   register number (0-31)
+ * @param[in] shift shift amount (0-3) * 16-bits
+ * @param[in] value immediate value
+ */
+static uint32_t mov(unsigned int reg, uint16_t shift, uint16_t value)
 {
-	struct ofi_intercept *intercept = data;
-	int phent, ret;
+	return (0x1a5 << 23) + ((uint32_t) shift << 21) + ((uint32_t) value << 5) + reg;
+}
 
-	phent = getauxval(AT_PHENT);
-	if (phent <= 0) {
-		FI_DBG(&core_prov, FI_LOG_MR, "failed to read phent size");
-		return -FI_EINVAL;
-	}
+/**
+ * @brief Generate a mov immediate with keep instruction
+ *
+ * @param[in] reg   register number (0-31)
+ * @param[in] shift shift amount (0-3) * 16-bits
+ * @param[in] value immediate value
+ */
+static uint32_t movk(unsigned int reg, uint16_t shift, uint16_t value)
+{
+	return (0x1e5 << 23) + ((uint32_t) shift << 21) + ((uint32_t) value << 5) + reg;
+}
 
-	ret = ofi_restore_dl_calls(info->dlpi_addr, info->dlpi_phdr,
-				   info->dlpi_name, info->dlpi_phnum,
-				   phent, intercept);
-	return ret;
+/**
+ * @brief Generate a branch to register instruction
+ *
+ * @param[in] reg   register number (0-31)
+ */
+static uint32_t br(unsigned int reg)
+{
+	return (0xd61f << 16) + (reg << 5);
 }
 
-static void ofi_restore_intercepts(void)
+static int
+PatchLoadImm(uintptr_t addr, unsigned int reg, uint64_t value)
 {
-	struct ofi_intercept *intercept;
+	*(uint32_t *) (addr +  0) = mov(reg, 3, value >> 48);
+	*(uint32_t *) (addr +  4) = movk(reg, 2, value >> 32);
+	*(uint32_t *) (addr +  8) = movk(reg, 1, value >> 16);
+	*(uint32_t *) (addr + 12) = movk(reg, 0, value);
+	return 16;
+}
```

Comment:
I was going to ask to rename this, but can we just merge it into the 1 calling place?

Reply:
Sure, updated.